[FBX - WIP] Various improvements for blender FBX specifically.

RevoluPowered opened the pull request:
Added a sanity check test, because I saw a bug on my PC which seems very rare, can be removed when this pr is ready I don't mind.

Fixes:
- normal data is now working from blender fbx files
- skinned meshes work more of the time from blender.
- blender fbx's *only* will reparent the mesh to the skeleton.
- blender has no support for pivots in an fbx.
- kenny character model skin is fixed.

:) 🍰 
<img width="1133" alt="Screenshot 2021-02-28 at 18 33 26" src="https://user-images.githubusercontent.com/748770/109431905-b1697800-7a00-11eb-9b84-4cf42e38ac19.png">

TODO:
- Fix animation time basis
- Remove pivots for blender fbx keyframes 
- default keys for blender fbx tracks probably need disabled

Bugs in engine:
- aabb's of skinned meshes are of the mesh without the skin (orientation then for the aabb is wrong, but maybe we can fix this?)

RevoluPowered committed:
Various improvements for blender FBX specifically.

Added a sanity check test, because I saw a bug on my PC which seems very rare.

fire commented:
If you can provide a sample project, my test plan is to run the project and do a visual inspection and then approve the pr.

akien-mga commented on tests/test_compiler_sanity.h:
```diff
@@ -0,0 +1,64 @@
+#ifndef TEST_SANITY_H
+#define TEST_SANITY_H
+
+#include "core/os/os.h"
+#include "core/string/ustring.h"
+#include "scene/resources/material.h"
```
Those headers don't seem used (at least `material.h`, didn't check that it compiles without though).

But I'd suggest moving this to a different PR as it's not related to FBX import (and that PR might need more details on what this test is for and how to reproduce the compiler issue it's trying to detect).

RevoluPowered replied:
ok, it was more for my own education as I had some bugs locally and some weird git behaviour gave me outdated modifications to files for some bizarre reason :)

I'll remove this

RevoluPowered committed:
fix regression with some orientations

RevoluPowered commented:
> If you can provide a sample project, my test plan is to run the project and do a visual inspection and then approve the pr.

I'll see if I can provide test files, but it really depends when I can.

I'm using kenny's paid for variant to test the skin fixes specifically.

I'll be working on keyframes fixes for blender next

RevoluPowered committed:
FIX: tracks in all importers off by default

RevoluPowered committed:
add has_named_bind

RevoluPowered committed:
fix blender keyframes not working

NOTE:
- still broken with orientation because blender doesn't know about FBX default track values.

RevoluPowered committed:
pre-rotation is breaking blender fbx

RevoluPowered committed:
fix bone comparison bug causing _1 everywhere

RevoluPowered committed:
remove pointless keyframes

RevoluPowered commented:
- 🍰 keyframes are fixed.
- it wont sample too many keyframes if there is no movement
- time basis is fixed
- i had to remove from to rotation for blender to work
- added flag to set the track to be an imported non-editable track

todo:
- fix compat bug with skin formula for maya fbx

akien-mga commented on modules/fbx/data/fbx_skeleton.cpp:
```diff
@@ -69,7 +69,7 @@ void FBXSkeleton::init_skeleton(const ImportState &state) {
 			const String bone_name = bone->bone_name;
 			int same_name_count = 0;
-			for (int y = x; y < skeleton_bone_count; y++) {
+			for (int y = x + 1; y < skeleton_bone_count; y++) {
```
This fixes #43820 right?

RevoluPowered replied:
yep

BastiaanOlij replied:
yeah we already merged that in 3 I think

akien-mga replied:
I extracted it in its own PR #47458 as this one is still WIP and will likely not make it in time for 3.3-stable.

akien-mga replied:
Ah you were right @BastiaanOlij, this has already been fixed in `3.x` via 9c296bcb59c. Should really have gone to `master` and be cherry-picked instead :)

BastiaanOlij replied:
yeah when I did that one Gordon said he was already including it in one of his fixes so :)

RevoluPowered committed:
reparenting is blender only

akien-mga commented:
Superseded by #59691 and #59653.
